### PR TITLE
blog: type tailoring (the `trivial` package)

### PR DESCRIPTION
**Pull request by bennn**
Blog post on type tailoring:
- the idea = you should let library authors program the type elaborator
- the example = counting groups in regex patterns
- quick comparison to Stephanie's keynote at popl 2017
- shouts out to our (rejected) paper

Two questions:
- Stephanie's keynote was a motivation for writing this, but on second thought I'm not sure the section about it fits
- The racket code isn't getting pygments formatting. What's the easiest way to fix that? (Should i convert the post to markdown?)

cc @mfelleisen @stchang

### CODE REVIEW

**Comment by mbutterick** [comment]
> The racket code isn't getting pygments formatting. What's the easiest way to fix that? 


Possibly:

https://github.com/greghendershott/frog/commit/b14c34c2a0357d180c7157f3b8dc0d7c17529935 <https://github.com/greghendershott/frog/commit/b14c34c2a0357d180c7157f3b8dc0d7c17529935>

File: blog/_src/posts/2017-02-26-type-tailoring.scrbl [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+ expand from source code to a typed intermediate representation.
+Using the @hyperlink[tsam-url]{@emph{Type Systems as Macros}} approach to building
+ typed language, the implementation is also straightforward, and the new typing
+ rules can be proven sound relative to the host language.
+
+<!-- more -->
+
+@section*{Background: Programming the Type Elaborator}
+
+Most typed languages include both a type checker and a @emph{type elaborator}.
```
Comment (samth) [inline blog/_src/posts/2017-02-26-type-tailoring.scrbl]: I don't think this is true about "most", regardless of how you draw the boundary.

File: blog/_src/posts/2017-02-26-type-tailoring.scrbl [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+The type elaborator translates source code to an
+ @hyperlink[mh-popl-1988]{explicitly typed} representation for the type
+ checker to validate.
+@; Good things including type inference happen during elaboration.
+Normally, programmers cannot extend the behavior of the type elaborator
+ without modifying the compiler.
+
+In @hyperlink[tr]{Typed Racket}, the Racket macro expander serves as a type
+ elaborator.
+The expander cooperates with forms such as Typed Racket's @racket[:] to
```
Comment (samth) [inline blog/_src/posts/2017-02-26-type-tailoring.scrbl]: This is not how `:` works.

File: blog/_src/posts/2017-02-26-type-tailoring.scrbl [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+In the cases, @racket[smart-match] defaults to Typed Racket's
+ @racket[regexp-match].
+Third, it fails to tailor calls where the pattern is a variable.
+
+The first problem is easy to fix by changing @racket[count-groups] to return
+ a list of @racket[num-groups] booleans encoding whether each capture group
+ definitely or maybe captures a string when the overall match is successful.
+The second problem is also just a matter of engineering.
+@margin-note{
+  The @|trivial-pkg| package includes a tailored @racket[regexp-match] that
```
Comment (samth) [inline blog/_src/posts/2017-02-26-type-tailoring.scrbl]: This involves changing the meaning of `define`, right?

Comment (samth) [inline blog/_src/posts/2017-02-26-type-tailoring.scrbl]: Also, I think you should advertise the pkg earlier in this post.

**Comment by bennn** [comment]
Fixed the formatting. LGTM.

**Comment by bennn** [comment]
ping @stamourv , can we post this?

**Comment by stamourv** [comment]
Sure.

**Comment by jackfirth** [comment]
Is this waiting on anything?

**Comment by stamourv** [comment]
I guess not. Merged.

**Comment by mbutterick** [comment]
Code blocks are inside `pre` tags, so they need to be hard-wrapped to a reasonable width (haven’t tested precisely but ~60 chars seems about right). The reason the [layout here](https://blog.racket-lang.org/2017/04/type-tailoring.html) is exploding is because two of the code boxes — the one under `regexp-match in Typed Racket` and the other under `Appendix: count-groups` — have long lines in them.

**Comment by bennn** [comment]
ok, I'll submit a PR to hard-wrap

**Pull request closed without merging**